registry/modules/stripe-subscriptions/controllers/subscription.ts

```diff
@@ -0,0 +1,354 @@
+
+import {
+  UserSubscription,
+  UserSubscriptionRepository,
+} from "@/repositories/subscription.interface.js";
+import { UserRepository } from "@/repositories/user.interface.js";
+
+import {
+  subscriptionNotFound,
+  noEmptySeatsToRemove,
+  noAvailableSeats,
+  cantRemoveSubOwner,
+} from "@/modules/stripe-subscriptions/utils/errors/subscriptions.js";
+
+const CHECKOUT_SUCCESS_URL = process.env.CHECKOUT_SUCCESS_URL as string;
+const CHECKOUT_CANCEL_URL = process.env.CHECKOUT_CANCEL_URL as string;
+
+type SubscriptionPrices = {
+  currency: string;
+  amount: number | null;
+};
+
+export type Subscription = {
+  plan: string;
+  priceId: string;
+  name: string;
+  description: string | null;
+  interval: "day" | "week" | "month" | "year";
+  price: SubscriptionPrices;
+  priceType: "per_unit" | "tiered";
+  features?: string[];
+};
+
+type UserSubsOutput = {
+  plan: string;
+  subscriptionId: string;
+  isOwner: boolean;
+  seats?: number;
+  createdAt: string;
+};
+
+interface SubscriptionController {
+  getSubscriptions: () => Promise<Subscription[]>;
+
+  getUserSubs: (
+    props: GetUserSubsSchema
+  ) => Promise<{ userSubscriptions: UserSubsOutput[] }>;
+
+  createCheckout: (props: CreateCheckoutSchema) => Promise<{ url: string }>;
+
+  createPaymentLink: (props: CreateCheckoutSchema) => Promise<{ url: string }>;
+
+  addUserToSeat: (props: AddUserToSeatSchema) => void;
+
+  updatePlan: (props: UpdatePlanSchema) => Promise<UserSubsOutput>;
+
+  updateSeats: (props: UpdateSubscriptionSeatsSchema) => void;
+
+  removeUserFromSeat: (props: RemoveUserFromSeatSchema) => void;
+
+  cancelSubscription: (props: CancelSubscriptionSchema) => void;
+
+  stopCancellation: (props: CancelSubscriptionSchema) => void;
+}
+
+export const createSubscriptionController = (
+  stripe: Stripe,
+  userSubRepository: UserSubscriptionRepository,
+  userRepository: UserRepository
+): SubscriptionController => {
+  const generatePlanKey = (priceKey: string | null, productName: string) => {
+    return priceKey || productName.toLowerCase().replaceAll(" ", "_");
+  };
+
+  return {
+    async getSubscriptions() {
+      const stripePrices = await stripe.prices.list({
+        active: true,
+        type: "recurring",
+        expand: ["data.product"],
+      });
+
+      const subscriptions: Subscription[] = [];
+
+      stripePrices.data.forEach((price) => {
+        const product = price.product as Stripe.Product; // We can assert the type as we expand the product object
+        const { name, description, marketing_features } = product;
+
+        const planKey = generatePlanKey(price.lookup_key, name);
+
+        const productFeatures = marketing_features
+          .filter((el) => el.name !== undefined)
+          .map((el) => el.name!);
+
+        subscriptions.push({
+          plan: planKey,
+          name,
+          description,
+          priceId: price.id,
+          interval: price.recurring!.interval,
```

Comment:
Why `!`? This mean Subscription is not properly typed.

Reply:
`Price.recurring` is optional if the price is not for a subscription. To create a subscription the user has to add a recurring property to Price on the Stripe Dashboard. Because here we handle subscriptions we know it will exist